### ISSUE TITLE
Unexpected JSC_TYPE_MISMATCH error with inferred type in removeChild

### ISSUE THREAD

**Issue by ricea** [google/closure-compiler#2382]
When I compile the following code

```javascript
// ==ClosureCompiler==
// @compilation_level ADVANCED_OPTIMIZATIONS
// @output_file_name default.js
// @formatting pretty_print
// ==/ClosureCompiler==

// ADD YOUR CODE HERE
let source;
function v() {
  const me = document.createElement('div');

  function setSource() {
    if (source) {
      me.removeChild(source); // line 8
      source = undefined;
    }
    source = document.createElement('div');
  }
}
```

with the online closure compiler, I get the following error:

```
JSC_TYPE_MISMATCH: actual parameter 1 of Node.prototype.removeChild does not match formal parameter
found   : (Element|undefined)
required: (Node|null) at line 8 character 23
        me.removeChild(source);
```

Explicitly specifying `@type {Element|undefined}` for `source` makes the error go away. Thanks to owler on StackOverflow for pointing this out: http://stackoverflow.com/questions/42858123/unexpected-jsc-type-mismatch-error-with-removechild/42863513#42863513

**Comment by dimvar** [comment]
Generally, the compiler can make fewer assumptions for variables used across scopes, so it's a good idea to declare their types.

**Comment by brad4d** [comment]
This is working as intended.
